File: leetcode/js/337.house-robber-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function (root) {
    let res = _rob(root);

    return Math.max(res.first, res.second);
};

/**
 * 
 * @param root 
 * @returns first: max_value can with root / max_value this root can get
 *          second: max_value  without root
 */
function _rob(root) {
    if (!root) {
        return { first: 0, second: 0 };
    }

    let left = _rob(root.left);
    let right = _rob(root.right);

    let _with = Math.max(root.val + left.second + right.second, left.first + right.first);
    let _without = left.first + right.first;

    return {
        first: Math.max(_with, _without),
        second: _without
    };
}

File: leetcode/cpp/337.house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include "oj_header.h"

class Solution
{
public:
    int rob(TreeNode *root)
    {
        auto res = _rob(root);

        return res.first;
    }

    pair<int, int> _rob(TreeNode *root)
    {
        if (!root)
        {
            return make_pair(0, 0);
        }

        auto left = _rob(root->left);
        auto right = _rob(root->right);
        // max value that root must not exist.
        int non_root_val = max(left.first, left.second) + max(right.first, right.second);
        // max value that root can in , and ** can not in**.
        int can_have_val = max(root->val + left.second + right.second, non_root_val);

        return make_pair(can_have_val, non_root_val);
    }
};
